### ISSUE TITLE
Typescript modules not supported properly when using jest to mock components

### ISSUE THREAD

**Issue by joelclimbsthings** [woocommerce/woocommerce-admin#7218]
**Describe the bug**
While attempting to write a test or two that stubs the response from `useSlot` in the `@woocommerce/experimental` package, I was consistently receiving this error:

```
  ● Test suite failed to run

    TypeError: Cannot read property 'TaskItem' of undefined

      38 | export { ExperimentalCollapsibleList as CollapsibleList } from './experimental-list/collapsible-list';
      39 | export { TaskItem } from './experimental-list/task-item';
    > 40 | export * from './inbox-note';
         |                      ^
```

This is imported using `jest.requireActual()` in the mock itself. The component that seems to be missing (task-item) is a `.tsx` file, and if I create an alternative `task-item.js` file (instead of `task-item.tsx`), then it runs properly with  no error.

**To Reproduce**

1. Create a test file
2. Add this block to mock out the `@woocommerce/experimental` package:
```js
jest.mock( '@woocommerce/experimental', () => {
	// Require the original module to not be mocked...
	const originalModule = jest.requireActual( '@woocommerce/experimental' );

	return {
		__esModule: true, // Use it when dealing with esModules
		...originalModule,
		useSlot: jest.fn().mockReturnValue( {
			fills: [],
		} ),
	};
} );
```

3. Observe the error.
4. Create a `task-item.js` file sibling to the `task-list.tsx` file, with any valid JSX component content.
```js
export const TaskItem = () => <p>Fake task item</p>
```
5. Notice the error disappears.

**Expected behavior**
The module is stubbed properly, with `useSlot` being mocked as expected.

**Comment by samueljseay** [comment]
@joelclimbsthings I can't reproduce this, It works fine for me with the code you provided. Which package are you putting the test in?

**Comment by joelclimbsthings** [comment]
Thanks for looking at this @samueljseay ! 

It was actually the `list.js` file created in this [as-yet-unmerged PR](https://github.com/woocommerce/woocommerce-admin/pull/7201), and I'm still able to reproduce it by pasting the above block into that file. You can clone that branch to attempt to reproduce.

> what command are you using to run the test

`npm run test:watch` & `npm run test`

> Did you ensure there had been an npm install

Yup, just double checked but no issues with dependencies. All other tests succeed just fine, with the exception of that test suite if the above block of code is added: 

![image](https://user-images.githubusercontent.com/444632/123181860-09f6d200-d443-11eb-8c33-22018cc36a5d.png)

**Comment by samueljseay** [comment]
Ah right you're creating the test in client! thanks I'll look further.

**Comment by samueljseay** [comment]
I did manage to reproduce this. It *might* be a related to an issue with re-exports and jest. I went down a rabbit hole of updating jest and installing ts-jest to see if I could fix it but alas it was not a simple fix. As far as I can tell though it seems babel-jest and ts-jest suffer this problem due to it being an upstream problem in Jest runtime.

A place to start might be just to update Jest to 27 and stay with babel-jest and then see if its still an issue.
